### PR TITLE
New format for proposal settings

### PR DESCRIPTION
**Pull request by joseivanlopez**
PBI: https://trello.com/c/I9K56rmD/566-5-allow-to-propose-a-data-partition-changes-in-the-proposal-logic

### CODE REVIEW

File: doc/old_and_new_proposal.md [hunk 1]
```diff
@@ -264,12 +264,12 @@ use case.
...
-    <lvm config:type="boolean">false</default_lvm>
-    <encrypt config:type="boolean">false</default_encrypt>
+    <lvm config:type="boolean">false</lvm>
+    <encrypt config:type="boolean">false</encrypt>
     <windows_delete_mode>all</windows_delete_mode>
-    <linux_delete>ondemand</linux_delete_mode>
+    <linux_delete_mode>ondemand</linux_delete_mode>
     <other_delete_mode>ondemand</other_delete_mode>
-    <lvm_vg_strategy_mode>use_available</lvm_vg_strategy>
+    <lvm_vg_strategy_mode>use_available</lvm_vg_strategy_mode>
```
Comment (ancorgs) [inline doc/old_and_new_proposal.md]: Actually the fix would be the other way around. The correct one is `lvm_vg_strategy` (without the "mode" part)

Comment (joseivanlopez) [reply doc/old_and_new_proposal.md]: Ohh, you are right :)

File: src/lib/y2storage/partitioning_features.rb [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    # When :to parameter is not indicated, the last value of feature path is considered
+    # the attribute name:
+    #
+    # @example
+    #   settings.lvm #=> nil
+    #   settings.load_feature(:proposal, :lvm)
+    #   settings.lvm #=> false
+    #
+    # When the feature needs to be loaded from a specific subsection, the subsection can
+    # be passed as :source parameter:
```
Comment (ancorgs) [inline src/lib/y2storage/partitioning_features.rb]: I got confused by the combination of this text and the example. From the text I was expecting this argument to be the name of the subsection, not its whole content.

Maybe rephrase it like:

By default, the feature is loaded from the content of the `<partitioning>` section of the product features. The argument `source` can be used to provide the content of any other section or subsection.

Comment (joseivanlopez) [reply src/lib/y2storage/partitioning_features.rb]: Yes, much better.

File: src/lib/y2storage/proposal_settings.rb [hunk 3]
```diff
@@ -24,58 +22,113 @@
...
+    attr_accessor :resize_windows
+
+    # @note :legacy format
+    # @return [DiskSize] root size used when calculating the :min size for
+    #   the proposal.
+    attr_accessor :root_base_size
+
+    # @note :legacy format
+    # @return [DiskSize] root size used when calculating the :desired size for
+    #   the proposal.
```
Comment (ancorgs) [inline src/lib/y2storage/proposal_settings.rb]: Although it's true that this is sometimes currently used as initial size for `:desired` (because we don't have a better alternative). That's not the real meaning of the attribute. This is the maximum size.

File: src/lib/y2storage/proposal_settings.rb [hunk 4]
```diff
@@ -24,58 +22,113 @@
...
+    # @return [Numeric] used to adjust size when distributing extra space
+    attr_accessor :root_space_percent
+
+    # @note :legacy format
+    # @return [Numeric] used to adjust size when using snapshots
+    attr_accessor :btrfs_increase_percentage
+
+    # @note :legacy format
+    # @return [DiskSize] min disk size to allow a separate home
+    attr_accessor :min_size_to_use_separate_home
```
Comment (ancorgs) [inline src/lib/y2storage/proposal_settings.rb]: Wow! Sure we can't come with a sorter name?

File: src/lib/y2storage/proposal_settings.rb [hunk 5]
```diff
@@ -24,58 +22,113 @@
...
+
+    # @note :legacy format
+    # @return [String] default btrfs subvolume path
+    attr_accessor :btrfs_default_subvolume
+
+    # @note :legacy format
```
Comment (ancorgs) [inline src/lib/y2storage/proposal_settings.rb]: In fact, this is not a proposal setting. This setting is used by the installer at the end of installation, but it doesn't affect the proposal.

File: src/lib/y2storage/proposal_settings.rb (hunk 4 above)
Comment (ancorgs) [reply src/lib/y2storage/proposal_settings.rb]: Nervermind, I just realized it was already called like that. You only documented it

Comment (joseivanlopez) [reply src/lib/y2storage/proposal_settings.rb]: Yes, I only reused it. In fact, right now this setting is not used at all. Do you have any alternative name in mind?

File: src/lib/y2storage/proposal_settings.rb (hunk 5 above)
Comment (ancorgs) [reply src/lib/y2storage/proposal_settings.rb]: So it's not used at all in the current `Y2Storage` and we should take the opportunity to remove it.

File: src/lib/y2storage/proposal_settings.rb [hunk 6]
```diff
@@ -24,58 +22,113 @@
...
+    # @return [DiskSize] home size used when calculating the :min size for
+    #   the proposal. If space is tight, {root_base_size} is used instead.
+    #   See also {Y2Storage::DevicesPlannerStrategies::Legacy#home_device}.
+    attr_accessor :home_min_size
+
+    # @note :legacy format
+    # @return [DiskSize] home size used when calculating the :desired size for
+    #   the proposal
```
Comment (ancorgs) [inline src/lib/y2storage/proposal_settings.rb]: Again, this is not accurate. By definition this is the maximum, not the initial for `:desired`

File: src/lib/y2storage/proposal_settings.rb [hunk 7]
```diff
@@ -24,58 +22,113 @@
...
     # @return [Boolean] whether to enlarge swap based on the RAM size, to ensure
     #   the classic suspend-to-ram works
     attr_accessor :enlarge_swap_for_suspend
 
+    # @note :legacy format
     # @return [String] device name of the disk in which / must be placed. If set
     #   to nil, the proposal will try to find a good candidate
     attr_accessor :root_device
 
+    # @note :legacy format
```
Comment (ancorgs) [inline src/lib/y2storage/proposal_settings.rb]: This is also used in the `:ng` format

File: src/lib/y2storage/proposal_settings.rb [hunk 8]
```diff
@@ -84,32 +137,74 @@ class ProposalUserSettings
...
-      @windows_delete_mode      = :ondemand
-      @linux_delete_mode        = :ondemand
-      @other_delete_mode        = :ondemand
+    attr_reader :other_delete_mode
+
+    # @note :ng format
+    # @return [Symbol] if the user decides to use LVM, strategy to decide the size
+    #   of the volume group (and, thus, the number and size of created physical volumes).
+    #
+    #   * :use_available The VG will be created to use all the available space, thus the
```
Comment (ancorgs) [inline src/lib/y2storage/proposal_settings.rb]: NP: Details in the yardoc formatting:

1)  `:use_available` vs `use_needed:` inconsistent placing of the colon.
2) For long descriptions that are not a one-liner we usually do

```ruby
# Description
#
# @note bla
#
# @return [Symbol]
attr_reader :whatever
```

Instead of that syntax with the `@return` on top.

File: src/lib/y2storage/proposal_settings.rb [hunk 9]
```diff
@@ -84,32 +137,74 @@ class ProposalUserSettings
...
+    end
+
+    # Current format used in control file
+    #
+    # Format is considered :ng only if subsections 'proposal' and 'volumes' are
+    # present in the 'partitioning' section.
+    #
+    # @return [Symbol] :ng or :legacy
+    def format
+      ng_format? ? NG_FORMAT : LEGACY_FORMAT
```
Comment (ancorgs) [inline src/lib/y2storage/proposal_settings.rb]: I see a problem with the design of `#format`, `#ng_format?` and so on. `ProductFeatures` is not immutable. Quite the opposite, nowadays looks like we are introducing more and more ways of changing it. So the mode should not be determined dynamically based on its current content. It should be determined during `#load_features` and stored in the `ProductFeatures` object alongside the read features.

File: src/lib/y2storage/proposal_settings.rb [hunk 10]
```diff
@@ -142,204 +237,185 @@ def delete_forced(type)
...
-        "  root_max_size: #{root_max_size}\n" \
-        "  root_space_percent: #{root_space_percent}\n" \
-        "  btrfs_increase_percentage: #{btrfs_increase_percentage}\n" \
-        "  min_size_to_use_separate_home: #{min_size_to_use_separate_home}\n" \
-        "  btrfs_default_subvolume: #{btrfs_default_subvolume}\n" \
-        "  root_subvolume_read_only: #{root_subvolume_read_only}\n" \
-        "  home_min_size: #{home_min_size}\n" \
-        "  home_max_size: #{home_max_size}\n" \
-        "  subvolumes: \n#{subvolumes}\n"
+    def apply_ng_defaults
```
Comment (ancorgs) [inline src/lib/y2storage/proposal_settings.rb]: NP: why are the right sides of the assignation to far away from the left sides?

File: src/lib/y2storage/proposal_settings.rb (hunk 9 above)
Comment (joseivanlopez) [reply src/lib/y2storage/proposal_settings.rb]: Stored in the `ProductFeatures`? Do you mean `ProposalSettings`?

File: src/lib/y2storage/proposal_settings.rb [hunk 11]
```diff
@@ -142,204 +237,185 @@ def delete_forced(type)
...
+      self.lvm                           ||= false
+      self.resize_windows                ||= true
+      self.windows_delete_mode           ||= :ondemand
+      self.linux_delete_mode             ||= :ondemand
+      self.other_delete_mode             ||= :ondemand
+      self.lvm_vg_strategy               ||= :use_needed
```
Comment (ancorgs) [inline src/lib/y2storage/proposal_settings.rb]: So far this is the only implemented strategy. But in the future I believe it should not be the default, because I'm pretty sure it's going to be by far the slowest one and because is not what we currently do in SLE or openSUSE. `:use_available` would be a more sensible default, I would say.

Comment (joseivanlopez) [reply src/lib/y2storage/proposal_settings.rb]: Maybe a #NOTE or #FIXME comment?

File: src/lib/y2storage/volume_specification.rb [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+
+    alias_method :proposed?, :proposed
+    alias_method :proposed_configurable?, :proposed_configurable
+    alias_method :adjust_by_ram?, :adjust_by_ram
+    alias_method :adjust_by_ram_configurable?, :adjust_by_ram_configurable
+    alias_method :snapshots?, :snapshots
+    alias_method :snapshots_configurable?, :snapshots_configurable
+
+    def initialize(volume_features)
+      @volume_features = volume_features
```
Comment (ancorgs) [inline src/lib/y2storage/volume_specification.rb]: Does this mean we store in memory everything twice (processed and raw)? Is there a practical reason or do we do it "just in case"?

File: src/lib/y2storage/volume_specification.rb [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+    def root?
+      mount_point && mount_point == "/"
+    end
+
+    def apply_fallbacks
+      apply_subvolumes_fallback
+      apply_fs_types_fallback
+    end
+
+    def apply_subvolumes_fallback
```
Comment (ancorgs) [inline src/lib/y2storage/volume_specification.rb]: Mmm, this makes impossible to specify that you want an empty list of subvolumes for root. Is that what we want?

File: src/lib/y2storage/volume_specification.rb [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+    def apply_fs_types_fallback
+      return unless fs_types.empty?
+
+      types = case mount_point
+      when "/"
+        Filesystems::Type.root_filesystems
+      when "/home"
+        Filesystems::Type.home_filesystems
+      else
+        []
```
Comment (ancorgs) [inline src/lib/y2storage/volume_specification.rb]: Maybe `[fs_type]` is a more sensible default here.

File: src/lib/y2storage/proposal_settings.rb [hunk 15]
```diff
@@ -142,204 +237,216 @@ def delete_forced(type)
...
+      ng_format? ? load_ng_features : load_legacy_features
     end
 
-  protected
+    # FIXME: Improve implementation. Use composition to encapsulate logic for
+    # ng and legacy formats
+    def apply_ng_defaults
+      self.lvm                 ||= false
+      self.resize_windows      ||= true
+      # TODO: once other strategies are available, change default value (:use_available)
```
Comment (ancorgs) [inline src/lib/y2storage/proposal_settings.rb]: This comment ended up in the wrong line. It refers to lvm_vg_strategy, not to windows_delete_mode.

The same happens again in the legacy version of the method
